### PR TITLE
[SPARK-33623][SQL] Add canDeleteWhere to SupportsDelete

### PR DESCRIPTION
**Pull request by aokolnychyi**
### What changes were proposed in this pull request?

This PR provides us with a way to check if a data source is going to reject the delete via `deleteWhere` at planning time.

### Why are the changes needed?

The only way to support delete statements right now is to implement ``SupportsDelete``. According to its Javadoc, that interface is meant for cases when we can delete data without much effort (e.g. like deleting a complete partition in a Hive table).

This PR actually provides us with a way to check if a data source is going to reject the delete via `deleteWhere` at planning time instead of just getting an exception during execution. In the future, we can use this functionality to decide whether Spark should rewrite this delete and execute a distributed query or it can just pass a set of filters.

Consider an example of a partitioned Hive table. If we have a delete predicate like `part_col = '2020'`, we can just drop the matching partition to satisfy this delete. In this case, the data source should return `true` from `canDeleteWhere` and use the filters it accepts in `deleteWhere` to drop the partition. I consider this as a delete without significant effort. At the same time, if we have a delete predicate like `id = 10`, Hive tables would not be able to execute this delete using a metadata only operation without rewriting files. In that case, the data source should return `false` from `canDeleteWhere` and we should use a more sophisticated row-level API to find out which records should be removed (the API is yet to be discussed, but we need this PR as a basis).

If we decide to support subqueries and all delete use cases by simply extending the existing API, this will mean all data sources will have to implement a lot of Spark logic to determine which records changed. I don't think we want to go that way as the Spark logic to determine which records should be deleted is independent of the underlying data source. So the assumption is that Spark will execute a plan to find which records must be deleted for data sources that return `false` from `canDeleteWhere`.
### Does this PR introduce _any_ user-facing change?

Yes but it is backward compatible. 

### How was this patch tested?

This PR comes with a new test.

### CODE REVIEW

**Comment by aokolnychyi** [comment]
cc @holdenk @dongjoon-hyun @dbtsai @rdblue @cloud-fan @viirya @sunchao

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java:39]: `without significant effort` looks like a misleading assumption here. Some Data source can implement this by themself with significant efforts due to its own reason.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java:47]: Shall we have `false` as a safer default?

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:1817]: Could you narrow down this test by removing irrelevant `PARTITIONED BY`?

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java
Comment (dongjoon-hyun) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java:47]: I'm wondering if there is a breaking change which we should have `true` here?

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java
Comment (rdblue) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java:39]: This comes from the documentation for `deleteWhere`:

> Implementations may reject a delete operation if the delete isn't possible without significant effort. For example, . . .

I think that phrasing is a bit more clear because it uses "implementations may reject", so what constitutes "significant effort" is determined by the implementation.

I think a clearer way to say it here is to refer to that standard: "Spark will call this method to check whether `deleteWhere` would reject the delete operation because it requires significant effort."

It would also help to have more context: this is for some sources to determine whether or not a metadata delete can be performed.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java
Comment (rdblue) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java:47]: Unfortunately, this would change the assumptions for existing implementations. Right now, if this interface is implemented, Spark will call `deleteWhere` for the delete. Returning `false` would cause Spark to skip it where the return of this method is used.

The original idea was to try to delete using `deleteWhere`, and if that fails to run a more expensive delete. But when we started implementing the more expensive delete, we needed to know during job planning, not job execution, whether the metadata-only delete can be done. This method solves that problem.

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Looks good overall.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java
Comment (viirya) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java:39]: > I think a clearer way to say it here is to refer to that standard: "Spark will call this method to check whether deleteWhere would reject the delete operation because it requires significant effort."

This sounds better as there is a standard between `canDeleteWhere` and `deleteWhere`.

So `canDeleteWhere` is a much light-weight approach to know `deleteWhere` will reject a delete operation without actually calling `deleteWhere`.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java
Comment (viirya) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java:49]: `deleteWhere` only possibly rejects a delete operation, even the delete isn't possible without significant effort. 

If `canDeleteWhere` returns false, does it mean `deleteWhere` definitely reject the delete operation?

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java
Comment (sunchao) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java:46]: Shall we clarify the relation between this and `deleteWhere`? say `deleteWhere` is only invoked when `canDeleteWhere` is positive. I must admit I'm somewhat confused when looking at this two and not sure what assumption should I make.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala
Comment (sunchao) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala:226]: is this exception handled later? the rewrite part for row deletion is a TBD?

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java
Comment (rdblue) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java:49]: I think that would be the case. In our use case, we can tell whether a delete is aligned with partitioning for this check. But, we can also scan through data to determine whether files themselves are fully matched (or not matched) by the filter. We would do the partitioning check here and the more expensive stats-based check in `deleteWhere`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala
Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala:226]: The rewrite would happen earlier. This just throws a good error message if `deleteWhere` will fail.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java
Comment (dongjoon-hyun) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java:47]: Got it, @rdblue .

**Comment by HyukjinKwon** [comment]
The API looks making sense to me too

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java
Comment (HyukjinKwon) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java:51]: Wait .. so does this basically for failing fast? What's the diff from throwing an exception `IllegalArgumentException` as documented here? We can still check first and fail fast as documented in this API.

Comment (cloud-fan) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java:51]: Yea I have the same question.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java
Comment (aokolnychyi) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java:47]: That's correct and the method returns `true` to keep the old behavior by default.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java
Comment (aokolnychyi) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java:39]: I agree with you @dongjoon-hyun @rdblue. I'll update the comment.

Comment (aokolnychyi) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java:39]: > So canDeleteWhere is a much light-weight approach to know deleteWhere will reject a delete operation without actually calling deleteWhere.

Yes, @viirya. It actually provides us with a way to check if a data source is going to reject the delete via `deleteWhere` at planning time instead of just getting an exception during execution. In the future, we can use this functionality to decide whether Spark should rewrite this delete and execute a distributed query or it can just pass a set of filters.

Consider an example of a partitioned Hive table. If we have a delete predicate like `part_col = '2020'`, we can just drop the matching partition to satisfy this delete. In this case, the data source should return `true` from `canDeleteWhere` and use the filters it accepts in `deleteWhere` to drop the partition. At the same time, if we have a delete predicate like `id = 10`, Hive tables would not be able to execute this delete using a metadata only operation without rewriting files. In that case, the data source should return `false` from `canDeleteWhere` and we should use a more sophisticated row-level API to find out which records should be removed (the API is yet to be discussed, but we need this PR as a basis).

If we decide to support subqueries and all delete use cases by simply extending the existing API, this will mean all data sources will have to implement a lot of Spark logic to determine which records changed. I don't think we want to go that way as the Spark logic to determine which records should be deleted is independent of the underlying data source. So the assumption is that Spark will execute a plan to find which records must be deleted for data sources that return `false` from `canDeleteWhere`.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java
Comment (aokolnychyi) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java:46]: Let me update the doc to make sure it is clear.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala
Comment (aokolnychyi) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala:226]: The rewrite part is yet to be done. This PR just adds a way to have more info at planning time.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java
Comment (aokolnychyi) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java:51]: I tried to elaborate on this [here](https://github.com/apache/spark/pull/30562#discussion_r533990021). Let me know if that makes sense, @HyukjinKwon @cloud-fan.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java
Comment (cloud-fan) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java:39]: My worry is that we may change the API back and forth if we don't have a clear big picture. Right now this patch is not useful as it only changes where we throw the exception, but I can see that this will be useful when we have the row-level delete API and we can use the `canDeleteWhere` to decide if we want to use the row-level API or not.

Do we have a sketch about how the row-level API would look like? It seems weird if we merge a patch that is useful to something that is not being designed yet.

Comment (HyukjinKwon) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java:39]: Mind filing JIRAs for the cases you'd leverage this API? I would prefer to have an explicit plan for that since three committers asked the same question. It's also best to move such description into the PR description.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java
Comment (aokolnychyi) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java:48]: Done. Could you check if it is descriptive enough, @sunchao?

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala
Comment (aokolnychyi) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala]: Done.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java
Comment (aokolnychyi) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java]: I've updated the PR description with more information from my comment above.

The design doc is being prepared. It felt to me that the community has enough confidence that we will need such a check at planning time (maybe, I misunderstand concepts and expectations behind `SupportsDelete`). In my view, it was pretty independent of the design of the row-level API and that's why I submitted it now.

Comment (aokolnychyi) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java]: I've updated the comment, @dongjoon-hyun @rdblue.

Comment (HyukjinKwon) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java]: I don't think simply planning time vs runtime justifies the API that dose the same thing that another API does. `SupportsDelete.deleteWhere` documents its purpose of failure cases which can fail fast as well (although it's runtime). 

Since it's planned to be used for something else as you elaborated, I think it should be best to make sure we all know the full context. We can think about other shapes of the API, for example, returning unhandled filters to make the error message to show which filters are not supported.

Comment (aokolnychyi) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java]: If we want to return a set of rejected filters to form a better message, that is OK for me. I am also not going to push for this change if there is no consensus. As I wrote before, it felt that the community had enough confidence that we would eventually need such a check at planning time based on how `SupportsDelete` is designed. If we want to delay this conversation until later, fine with me too.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. With @rdblue 's and @aokolnychyi 's explanation, the updated patch looks clear to me now. Thank you, @aokolnychyi and @rdblue .

Could you review once more, @holdenk @dbtsai @cloud-fan @viirya @sunchao ?

**Comment by dongjoon-hyun** [comment]
Also, cc @gatorsmile . Please let us know if you have some comments .

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java
Comment (rdblue) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java]: > My worry is that we may change the API back and forth if we don't have a clear big picture. Right now this patch is not useful as it only changes where we throw the exception, but I can see that this will be useful when we have the row-level delete API and we can use the canDeleteWhere to decide if we want to use the row-level API or not.

This is exactly the reason for adding this API. It is a step toward rewriting plans for row-level DELETE and MERGE operations. The current `deleteWhere` exception approach happens while running the physical plan, when it is too late to rewrite the plan for row-level changes. Adding the `canDeleteWhere` check fixes that problem.

Since you can easily see how it will be used, what is the concern about adding this?

Comment (cloud-fan) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java]: My concern is not about this PR itself, but about the order of commits. It seems more natural to have the row-level delete API first, and then do this change. At that time we can have tests to verify if we can switch correctly.

If you are working with @aokolnychyi on this feature, and you two think this is better for your development, please go ahead and merge it.

Comment (rdblue) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java]: Thanks, we do think that it is helpful to have it in this order. This was an early problem that we ran into and the solution is clear.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java
Comment (sunchao) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java:48]: Thanks for the update @aokolnychyi . I'm wondering if it makes sense to clearly point out that `canDeleteWhere` is called at planning time, say: "Spark will call this method **at planning time** to check if the delete is possible without significant effort."

Also for `deleteWhere`, I'm wondering if the following makes sense:
"Delete data from a data source table that matches filter expressions. **This is only invoked when the canDeleteWhere check succeed**".

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala
Comment (sunchao) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala:226]: I see, thanks. So this method will be called in an earlier place and before rewrite once the rewrite part is ready, is that right?

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java
Comment (aokolnychyi) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java:48]: Makes sense to me, @sunchao. Updated.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala
Comment (aokolnychyi) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala:226]: It is going to be called at planning time to check if we should apply the rewrite or just pass filters down.

**Comment by aokolnychyi** [comment]
I've updated the doc per @sunchao suggestion. Let me know if there are any other open questions.

**Review by sunchao — APPROVED** [review]
Thanks @aokolnychyi ! the updated doc looks great to me. I'm LTGM (non-binding) on this PR.

**Comment by aokolnychyi** [comment]
The recent test failures do not seem related to this PR.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java
Comment (HyukjinKwon) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java]: Can you guys just file related JIRAs for that before merging? That's all I asked.

**Comment by HyukjinKwon** [comment]
Hey, please don't ignore my opinion, and file related JIRAs (https://github.com/apache/spark/pull/30562#discussion_r534022263).

**Comment by dongjoon-hyun** [comment]
Of course, he will do, @HyukjinKwon . We are still under reviews. In addition to that, is this (https://github.com/apache/spark/pull/30562#discussion_r533898538) resolved?

**Comment by HyukjinKwon** [comment]
Thanks @dongjoon-hyun. Yes. It's okay to edit the JIRAs later, and I understand the design doc is in progress. I just wanted to make sure we all know what to do after this.

**Comment by aokolnychyi** [comment]
Filing JIRAs is something I will do for sure before this one is merged, @HyukjinKwon. I was referring to any open points related to this PR and whether we have enough consensus and everybody is ok with the change. If anybody still has unresolved concerns, let's discuss until we all agree.

**Comment by HyukjinKwon** [comment]
I am okay with this PR. I have no unresolved concerns except that I prefer to know and make sure we have a plan. Easiest way should be to file relevant JIRAs.

**Comment by aokolnychyi** [comment]
I've created SPARK-33642 as a parent JIRA and 3 substasks for DELETE/UPDATE/MERGE. The parent one is where the design doc should be and where the discussion should happen.

I don't see any open points on this PR now but I propose we wait a bit more for additional feedback to make sure everybody is on the same page.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java
Comment (aokolnychyi) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java:48]: This was resolved.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java
Comment (aokolnychyi) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java]: Just to update this thread, JIRAs were created. [Here](https://github.com/apache/spark/pull/30562#issuecomment-737779083) is a summary.

**Comment by dongjoon-hyun** [comment]
Thank you all! I'll merge this for Apache Spark 3.1.

**Pull request closed without merging**
